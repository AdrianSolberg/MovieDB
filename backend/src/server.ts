import { ApolloServer } from "@apollo/server";
import { startStandaloneServer } from "@apollo/server/standalone";
import depthLimit from "graphql-depth-limit";
import type { ListenOptions } from "net";
import connectToDatabase from "./db/database.js";
import resolvers from "./resolvers/resolvers.js";
import { typeDefs } from "./schemas/schema.js";

/**
 * Creates and starts an Apollo Server instance. Will connect to the database before starting the server.
 *
 * @param listenOptions - The options for the server to listen on. (e.g. port)
 * @param databaseUri - The URI of the database to connect to.
 * @returns An object containing the server instance and the URL it is listening on.
 *
 * @throws Will throw an error if the server fails to start or the database connection fails.
 */
export async function createApolloServer(
    listenOptions: ListenOptions,
    databaseUri: string
) {
    const server = new ApolloServer({
        typeDefs,
        resolvers,
        validationRules: [depthLimit(2)],
    });

    await connectToDatabase(databaseUri);

    const { url } = await startStandaloneServer(server, {
        listen: listenOptions,
    });

    return { server, url };
}
