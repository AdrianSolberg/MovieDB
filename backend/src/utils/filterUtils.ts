export type MovieFilters = {
    genre: string[];
    rating: string[];
    releaseYear: string[];
    language: string[];
    status: string[];
    runtime: string[];
};

const filterByGenres = (selectedGenres: string[]) => {
    return selectedGenres.length ? { genres: { $all: selectedGenres } } : {};
};

const filterByRating = (selectedRatings: string[]) => {
    if (selectedRatings.length) {
        const mappedRatings = selectedRatings.map(rating => parseInt(rating) * 2);
        const ratingFilters =  mappedRatings.map(rating => {
            return {
                vote_average: {
                    $gte: rating - 1,
                    $lt: rating + 1
                }
            };
        })
        return { $or: ratingFilters }
    }
    return {};
};

const filterByReleaseYear = (selectedDecades: string[]) => {
    if (selectedDecades.length) {
        const yearFilters = selectedDecades.map((decade) => {
            const startYear = parseInt(decade.slice(0, 4));
            return {
                release_date: {
                    $gte: new Date(`${startYear}-01-01`),
                    $lt: new Date(`${startYear + 10}-01-01`),
                },
            };
        });
        return { $or: yearFilters };
    }
    return {};
};

const filterByLanguage = (selectedLanguages: string[]) => {
    return selectedLanguages.length
        ? { spoken_languages: { $all: selectedLanguages } }
        : {};
};

const filterByStatus = (selectedStatuses: string[]) => {
    return selectedStatuses.length ? { status: { $in: selectedStatuses } } : {};
};

const filterByRuntime = (selectedRuntimes: string[]) => {
    if (selectedRuntimes.length) {
        const runtimeQueries = selectedRuntimes.map((runtimeFilter) => {
            switch (runtimeFilter) {
                case "Less than 1 hour":
                    return { runtime: { $lt: 60 } };
                case "1 - 2 hours":
                    return { runtime: { $gte: 60, $lt: 120 } };
                case "2 - 3 hours":
                    return { runtime: { $gte: 120, $lt: 180 } };
                case "3 hours or more":
                    return { runtime: { $gte: 180 } };
                default:
                    return {};
            }
        });
        return { $or: runtimeQueries }
    }
    return {};
};

export const createFilterQuery = (filters: MovieFilters) => {
    const queryConditions = [
        filterByGenres(filters.genre),
        filterByRating(filters.rating),
        filterByReleaseYear(filters.releaseYear),
        filterByLanguage(filters.language),
        filterByStatus(filters.status),
        filterByRuntime(filters.runtime),
    ].filter(condition => Object.keys(condition).length > 0);

    return queryConditions.length > 0 ? { $and: queryConditions } : {};
};