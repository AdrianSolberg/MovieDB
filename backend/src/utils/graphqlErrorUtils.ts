import { GraphQLError } from "graphql";

export function createBadUserInputError(message: string) {
    return new GraphQLError(message, {
        extensions: { code: "BAD_USER_INPUT" },
    });
}

export function validateSkipLimit(skip: number, limit: number) {
    if (limit == null || limit < 1) {
        return createBadUserInputError(
            "Limit must be an integer of size at least 1."
        );
    }
    if (limit > 100) {
        return createBadUserInputError(
            "Limit must be an integer of size at most 100."
        );
    }
    if (skip == null || skip < 0) {
        return createBadUserInputError(
            "Skip must be an integer of size at least 0."
        );
    }
    if (!Number.isInteger(skip) || !Number.isInteger(limit)) {
        return createBadUserInputError("Skip and limit must be integers.");
    }
    return null;
}

export function validateUsername(username: string) {
    const usernameMinLength = 3;
    const usernameMaxLength = 20;
    if (username.length < usernameMinLength) {
        return createBadUserInputError(
            `Username must be at least ${usernameMinLength} characters long.`
        );
    }
    if (username.length > 20) {
        return createBadUserInputError(
            `Username must be at most ${usernameMaxLength} characters long.`
        );
    }
    return null;
}

export function validateReview(comment: string) {
    const reviewMaxLength = 1500;
    if (comment.trim().length > reviewMaxLength) {
        return createBadUserInputError(
            `Review must be at most ${reviewMaxLength} characters.`
        );
    }
    return null;
}
