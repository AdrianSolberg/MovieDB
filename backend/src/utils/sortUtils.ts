import { SortOrder } from "mongoose";

export enum SortingType {
    NEWEST_FIRST = "NEWEST_FIRST",
    OLDEST_FIRST = "OLDEST_FIRST",
    BEST_RATED = "BEST_RATED",
    WORST_RATED = "WORST_RATED",
    LONGEST_RUNTIME = "LONGEST_RUNTIME",
    SHORTEST_RUNTIME = "SHORTEST_RUNTIME",
}

export const getSortOrder = (
    sortOption: SortingType
): { [key: string]: SortOrder } => {
    switch (sortOption) {
        case SortingType.NEWEST_FIRST:
            return { release_date: -1 };
        case SortingType.OLDEST_FIRST:
            return { release_date: 1 };
        case SortingType.BEST_RATED:
            return { vote_average: -1 };
        case SortingType.WORST_RATED:
            return { vote_average: 1 };
        case SortingType.LONGEST_RUNTIME:
            return { runtime: -1 };
        case SortingType.SHORTEST_RUNTIME:
            return { runtime: 1 };
        default:
            return {};
    }
};
