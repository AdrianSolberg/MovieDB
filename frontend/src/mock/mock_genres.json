[
    {
        "id": 0,
        "name": "Drama"
    },
    {
        "id": 1,
        "name": "Romance"
    },
    {
        "id": 2,
        "name": "Western"
    },
    {
        "id": 3,
        "name": "Animation"
    },
    {
        "id": 4,
        "name": "War"
    },
    {
        "id": 5,
        "name": "Science Fiction"
    },
    {
        "id": 6,
        "name": "Comedy"
    },
    {
        "id": 7,
        "name": "Fantasy"
    },
    {
        "id": 8,
        "name": "Adventure"
    },
    {
        "id": 9,
        "name": "Crime"
    },
    {
        "id": 10,
        "name": "Thriller"
    },
    {
        "id": 11,
        "name": "Family"
    },
    {
        "id": 12,
        "name": "Action"
    },
    {
        "id": 13,
        "name": "Mystery"
    }
]
