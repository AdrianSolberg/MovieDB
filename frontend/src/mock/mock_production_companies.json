[
    "Lynda Obst Productions",
    "Pascal Pictures",
    "Fairview Entertainment",
    "Silver Pictures",
    "Walt Disney Pictures",
    "Visiona Romantica",
    "DC Films",
    "Mandate International",
    "Patalex IV Productions Limited",
    "Amblin Entertainment",
    "Heyday Films",
    "Universal Pictures",
    "Bron Studios",
    "Juno Pix",
    "Color Force",
    "LStar Capital",
    "Miramax",
    "The Saul Zaentz Company",
    "Sikelia Productions",
    "Red Granite Pictures",
    "The Weinstein Company",
    "Knickerbocker Films",
    "Columbia Pictures",
    "Groucho II Film Partnership",
    "Syncopy",
    "P of A Productions Limited",
    "Zehnte Babelsberg Film",
    "The Linson Company",
    "Pixar",
    "Appian Way",
    "The Steve Tisch Company",
    "Genre Films",
    "Paramount",
    "Kennedy Miller Productions",
    "Lucasfilm Ltd.",
    "New Line Cinema",
    "Atman Entertainment",
    "Cecchi Gori Pictures",
    "Castle Rock Entertainment",
    "Isobel Griffiths",
    "Phoenix Pictures",
    "DC Entertainment",
    "Dune Entertainment",
    "20th Century Fox",
    "Marvel Studios",
    "Lightstorm Entertainment",
    "WingNut Films",
    "Fox 2000 Pictures",
    "RatPac Entertainment",
    "Lionsgate",
    "Warner Bros. Pictures",
    "Patalex III Productions Limited",
    "Jersey Films",
    "Joint Effort",
    "Village Roadshow Pictures",
    "EMJAG Productions",
    "The Donners' Company",
    "1492 Pictures",
    "A Band Apart",
    "DC Comics",
    "Atlas Entertainment",
    "Ingenious Media",
    "Legendary Pictures",
    "Taurus Film",
    "Wendy Finerman Productions",
    "Peninsula Films",
    "Lin Pictures",
    "Regency Enterprises",
    "Jerry Bruckheimer Films"
]
