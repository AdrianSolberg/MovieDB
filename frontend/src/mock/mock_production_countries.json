[
    {
        "id": 0,
        "name": "New Zealand"
    },
    {
        "id": 1,
        "name": "Canada"
    },
    {
        "id": 2,
        "name": "South Africa"
    },
    {
        "id": 3,
        "name": "United Kingdom"
    },
    {
        "id": 4,
        "name": "Germany"
    },
    {
        "id": 5,
        "name": "Australia"
    },
    {
        "id": 6,
        "name": "France"
    },
    {
        "id": 7,
        "name": "United States of America"
    }
]
