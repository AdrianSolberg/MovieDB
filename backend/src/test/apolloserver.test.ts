import { ApolloServer, BaseContext } from "@apollo/server";
import axios from "axios";
import { MongoMemoryServer } from "mongodb-memory-server";
import mongoose from "mongoose";
import {
    afterAll,
    afterEach,
    beforeAll,
    beforeEach,
    describe,
    expect,
    it,
} from "vitest";
import { createApolloServer } from "../server";
import MovieModel from "../models/movie.model";
import ReviewModel from "../models/review.model";
import { SortingType } from "../utils/sortUtils";
import { movies, reviews } from "./mock/util";
import {
    ADD_REVIEW,
    DELETE_REVIEW,
    GET_FILTERS,
    GET_LATEST_REVIEWS,
    GET_MOVIE,
    GET_MOVIES,
    GET_USER_REVIEWS,
} from "./queries";

const allVariables = {
    movie: {
        movieId: 157336,
    },
    movies: {
        skip: 0,
        limit: 3,
        filters: {
            Decade: [],
            Rating: [],
            Genre: [],
            Status: [],
            Runtime: [],
        },
        sortOption: SortingType.MOST_POPULAR,
        search: "",
    },
    filters: {
        appliedFilters: {
            Decade: [],
            Rating: [],
            Genre: ["Thriller"],
            Status: [],
            Runtime: [],
        },
        search: "the",
    },
    latestReviews: {
        limit: 2,
    },
    userReviews: {
        username: "testuser2",
    },
    addReview: {
        movieId: 565770,
        username: "testuser3",
        rating: 4,
        comment: "Cool movie",
    },
    deleteReview: {
        id: "673625f49c33bb841a3fef05",
    },
};

describe("resolvers", () => {
    let server: ApolloServer<BaseContext>;
    let url: string;
    let mongoServer: MongoMemoryServer;

    beforeAll(async () => {
        mongoServer = await MongoMemoryServer.create();
        const mongoUri = mongoServer.getUri();

        // createApolloServer handles connecting to the database
        ({ server, url } = await createApolloServer({ port: 0 }, mongoUri));

        await MovieModel.createCollection();
        await ReviewModel.createCollection();
    });

    beforeEach(async () => {
        await MovieModel.insertMany(movies);
        await ReviewModel.insertMany(reviews);
    });

    afterEach(async () => {
        await MovieModel.deleteMany({});
        await ReviewModel.deleteMany({});
    });

    afterAll(async () => {
        if (mongoServer) {
            await mongoServer.stop();
        }
        if (server) {
            await server.stop();
        }

        await mongoose.connection.close();
    });

    for (const [query, name, variables] of [
        [GET_MOVIE, "movie", allVariables.movie],
        [GET_MOVIES, "movies", allVariables.movies],
        [GET_FILTERS, "filters", allVariables.filters],
        [GET_LATEST_REVIEWS, "latestReviews", allVariables.latestReviews],
        [GET_USER_REVIEWS, "userReviews", allVariables.userReviews],
        [DELETE_REVIEW, "deleteReview", allVariables.deleteReview],
    ]) {
        it(`correctly handles ${name} resolver`, async () => {
            const response = await axios.post(url, { query, variables });
            expect(response.data.errors).toBeUndefined();
            expect(response.data.data).toMatchSnapshot();
        });
    }

    it("correctly handles addReview resolver", async () => {
        // Can't use toMatchSnapshot because the id is generated by the database
        const response = await axios.post(url, {
            query: ADD_REVIEW,
            variables: allVariables.addReview,
        });
        expect(response.data.errors).toBeUndefined();
        expect(response.data.data.addReview).toMatchObject({
            username: "testuser3",
            rating: 4,
            comment: "Cool movie",
        });
    });
});
