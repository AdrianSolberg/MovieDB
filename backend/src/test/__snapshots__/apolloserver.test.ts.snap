// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolvers > correctly handles deleteReview resolver 1`] = `
{
  "deleteReview": {
    "_id": "673625f49c33bb841a3fef05",
    "comment": "Great movie, but I can't talk about it. You know the rule.",
    "date": "2024-11-14T16:31:48.039Z",
    "movie": {
      "_id": 550,
      "title": "Fight Club",
      "vote_average": 8.438,
      "vote_count": 27238,
    },
    "rating": 5,
    "username": "testuser1",
  },
}
`;

exports[`resolvers > correctly handles filters resolver 1`] = `
{
  "filters": {
    "Decade": [
      {
        "hits": 5,
        "name": "2020s",
      },
      {
        "hits": 1,
        "name": "2010s",
      },
      {
        "hits": 0,
        "name": "1990s",
      },
    ],
    "Genre": [
      {
        "hits": 4,
        "name": "Action",
      },
      {
        "hits": 0,
        "name": "Adventure",
      },
      {
        "hits": 0,
        "name": "Animation",
      },
      {
        "hits": 0,
        "name": "Comedy",
      },
      {
        "hits": 1,
        "name": "Crime",
      },
      {
        "hits": 0,
        "name": "Drama",
      },
      {
        "hits": 0,
        "name": "Family",
      },
      {
        "hits": 0,
        "name": "Fantasy",
      },
      {
        "hits": 0,
        "name": "History",
      },
      {
        "hits": 2,
        "name": "Horror",
      },
      {
        "hits": 3,
        "name": "Mystery",
      },
      {
        "hits": 0,
        "name": "Romance",
      },
      {
        "hits": 1,
        "name": "Science Fiction",
      },
      {
        "hits": 6,
        "name": "Thriller",
      },
      {
        "hits": 0,
        "name": "War",
      },
    ],
    "Rating": [
      {
        "hits": 0,
        "name": "5",
      },
      {
        "hits": 2,
        "name": "4",
      },
      {
        "hits": 4,
        "name": "3",
      },
      {
        "hits": 0,
        "name": "2",
      },
      {
        "hits": 0,
        "name": "1",
      },
      {
        "hits": 0,
        "name": "0",
      },
    ],
    "Runtime": [
      {
        "hits": 0,
        "name": "Less than 1 hour",
      },
      {
        "hits": 4,
        "name": "1 - 2 hours",
      },
      {
        "hits": 2,
        "name": "2 - 3 hours",
      },
      {
        "hits": 0,
        "name": "3 hours or more",
      },
    ],
    "Status": [
      {
        "hits": 6,
        "name": "Released",
      },
      {
        "hits": 0,
        "name": "In Production",
      },
      {
        "hits": 0,
        "name": "Post Production",
      },
      {
        "hits": 0,
        "name": "Planned",
      },
    ],
  },
}
`;

exports[`resolvers > correctly handles latestReviews resolver 1`] = `
{
  "latestReviews": [
    {
      "_id": "673627c79c33bb841a3fef74",
      "comment": "This is going to be quite good.",
      "date": "2024-11-14T16:39:35.185Z",
      "movie": {
        "_id": 83533,
        "poster_path": "/eZXoA5BjUxm1fdAKb3lPvbDY2L8.jpg",
        "title": "Avatar 3",
      },
      "rating": 3,
      "username": "testuser1",
    },
    {
      "_id": "673625f49c33bb841a3fef05",
      "comment": "Great movie, but I can't talk about it. You know the rule.",
      "date": "2024-11-14T16:31:48.039Z",
      "movie": {
        "_id": 550,
        "poster_path": "/pB8BM7pdSp6B6Ih7QZ4DrQ3PmJK.jpg",
        "title": "Fight Club",
      },
      "rating": 5,
      "username": "testuser1",
    },
  ],
}
`;

exports[`resolvers > correctly handles movie resolver 1`] = `
{
  "movie": {
    "_id": 157336,
    "genres": [
      "Adventure",
      "Drama",
      "Science Fiction",
    ],
    "reviews": [
      {
        "_id": "6736255c9c33bb841a3fee9a",
        "comment": "It's quite good",
        "date": "2024-11-14T16:29:16.190Z",
        "rating": 4,
        "username": "testuser2",
      },
      {
        "_id": "6736254b9c33bb841a3fee95",
        "comment": "Best movie ever made!",
        "date": "2024-11-14T16:28:59.076Z",
        "rating": 5,
        "username": "testuser1",
      },
    ],
    "title": "Interstellar",
  },
}
`;

exports[`resolvers > correctly handles movies resolver 1`] = `
{
  "movies": [
    {
      "_id": 565770,
      "poster_path": "/mXLOHHc1Zeuwsl4xYKjKh2280oL.jpg",
      "release_date": "2023-08-16T00:00:00.000Z",
      "runtime": 128,
      "title": "Blue Beetle",
      "vote_average": 7.139,
    },
    {
      "_id": 980489,
      "poster_path": "/51tqzRtKMMZEYUpSYkrUE7v9ehm.jpg",
      "release_date": "2023-08-09T00:00:00.000Z",
      "runtime": 135,
      "title": "Gran Turismo",
      "vote_average": 8.064965909090908,
    },
    {
      "_id": 968051,
      "poster_path": "/c9kVD7W8CT5xe4O3hQ7bFWwk68U.jpg",
      "release_date": "2023-09-06T00:00:00.000Z",
      "runtime": 110,
      "title": "The Nun II",
      "vote_average": 6.551426630434783,
    },
  ],
}
`;

exports[`resolvers > correctly handles userReviews resolver 1`] = `
{
  "userReviews": [
    {
      "_id": "6736255c9c33bb841a3fee9a",
      "comment": "It's quite good",
      "date": "2024-11-14T16:29:16.190Z",
      "movie": {
        "_id": 157336,
        "poster_path": "/gEU2QniE6E77NI6lCU6MxlNBvIx.jpg",
        "title": "Interstellar",
      },
      "rating": 4,
    },
  ],
}
`;
